algorithms/src/snark/marlin/ahp/prover/round_functions/second.rs

```diff
@@ -15,360 +15,178 @@
-        DensePolynomial,
-        EvaluationDomain,
-    },
+    fft::{polynomial::PolyMultiplier, DensePolynomial, EvaluationDomain, Evaluations as EvaluationsOnDomain},
     polycommit::sonic_pc::{LabeledPolynomial, PolynomialInfo, PolynomialLabel},
     snark::marlin::{
-        ahp::{
-            indexer::{CircuitId, Matrix},
-            verifier,
-            AHPForR1CS,
-            UnnormalizedBivariateLagrangePoly,
-        },
+        ahp::{verifier, AHPForR1CS},
         prover,
+        witness_label,
+        Circuit,
+        CircuitId,
         MarlinMode,
     },
 };
-use snarkvm_fields::PrimeField;
-use snarkvm_utilities::{cfg_iter, cfg_iter_mut, ExecutionPool};
-
-use itertools::Itertools;
+use anyhow::ensure;
 use rand_core::RngCore;
+use snarkvm_fields::PrimeField;
+use snarkvm_utilities::{cfg_into_iter, cfg_iter_mut, cfg_reduce, ExecutionPool};
 
 #[cfg(not(feature = "serial"))]
 use rayon::prelude::*;
 
-struct LincheckPoly<F: PrimeField> {
-    circuit_id: CircuitId,
-    poly: DensePolynomial<F>,
-}
-
 impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
     /// Output the number of oracles sent by the prover in the second round.
-    pub fn num_second_round_oracles() -> usize {
-        2
+    pub const fn num_second_round_oracles() -> usize {
+        1
     }
 
-    /// Output the degree bounds of oracles in the first round.
-    pub fn second_round_polynomial_info(constraint_domain_size: usize) -> BTreeMap<PolynomialLabel, PolynomialInfo> {
-        [
-            PolynomialInfo::new("g_1".into(), Some(constraint_domain_size - 2), Self::zk_bound()),
-            PolynomialInfo::new("h_1".into(), None, None),
-        ]
-        .into_iter()
-        .map(|info| (info.label().into(), info))
-        .collect()
+    /// Output the degree bounds of oracles in the second round.
+    pub fn second_round_polynomial_info() -> BTreeMap<PolynomialLabel, PolynomialInfo> {
+        [PolynomialInfo::new("h_0".into(), None, None)].into_iter().map(|info| (info.label().into(), info)).collect()
     }
 
     /// Output the second round message and the next state.
     pub fn prover_second_round<'a, R: RngCore>(
         verifier_message: &verifier::FirstMessage<F>,
         mut state: prover::State<'a, F, MM>,
         _r: &mut R,
-    ) -> (prover::SecondOracles<F>, prover::State<'a, F, MM>) {
+    ) -> Result<(prover::SecondOracles<F>, prover::State<'a, F, MM>), anyhow::Error> {
         let round_time = start_timer!(|| "AHP::Prover::SecondRound");
 
         let zk_bound = Self::zk_bound();
 
         let max_constraint_domain = state.max_constraint_domain;
 
-        let verifier::FirstMessage { alpha, eta_b, eta_c, batch_combiners } = verifier_message;
+        let verifier::FirstMessage { batch_combiners, .. } = verifier_message;
 
-        let summed_z_m_and_t = Self::calculate_summed_z_m_and_t(&state, *alpha, *eta_b, *eta_c, batch_combiners);
+        let h_0 = Self::calculate_lhs_second(&mut state, batch_combiners)?;
```

Comment:
Should we give this a more informative name?

Reply:
Absolutely, took a stab: 46106cd8c
Witness seems generic enough to be consistent across rounds and for the next lookup PR.